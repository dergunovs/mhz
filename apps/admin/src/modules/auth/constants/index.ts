export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
