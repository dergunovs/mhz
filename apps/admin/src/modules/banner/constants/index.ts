export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
