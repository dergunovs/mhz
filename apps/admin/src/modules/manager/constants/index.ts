export const URL_MANAGER = '/managers';
