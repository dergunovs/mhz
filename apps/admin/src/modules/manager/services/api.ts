import { ComputedRef, Ref } from 'vue';

import { api, IPageQuery, convertParams } from 'mhz-helpers';
import { API_MANAGER, IBaseReply, IManager } from 'mhz-contracts';

export async function getManagersApi(query: Ref<IPageQuery | number>) {
  const params = convertParams(query);

  const { data } = await api.get<{ data: IManager[]; total: number }>(API_MANAGER, { params });

  return data;
}

export async function getManagerApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: IManager }>(`${API_MANAGER}/${id.value}`);

  return data.data;
}

export async function postManagerApi(formData: IManager) {
  const { data } = await api.post<IBaseReply>(API_MANAGER, formData);

  return data;
}

export async function updateManagerApi(formData: IManager) {
  const { data } = await api.patch<IBaseReply>(`${API_MANAGER}/${formData._id}`, formData);

  return data;
}

export async function deleteManagerApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_MANAGER}/${id}`);

  return data;
}
