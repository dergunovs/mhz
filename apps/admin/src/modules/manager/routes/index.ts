import { URL_MANAGER, URL_MANAGER_CREATE, URL_MANAGER_EDIT } from '@/manager/constants';

export const managerRoutes = [
  {
    path: URL_MANAGER,
    name: 'ManagerList',
    component: () => import('@/manager/pages/ManagerListPage.vue'),
  },
  {
    path: URL_MANAGER_CREATE,
    name: 'ManagerCreate',
    component: () => import('@/manager/pages/ManagerCreatePage.vue'),
  },
  {
    path: `${URL_MANAGER_EDIT}/:manager`,
    name: 'ManagerEdit',
    component: () => import('@/manager/pages/ManagerEditPage.vue'),
  },
];
