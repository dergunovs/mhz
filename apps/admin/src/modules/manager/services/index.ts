import { ComputedRef, Ref } from 'vue';

import { IManager } from 'mhz-types';
import { api, useMutation, useQuery, IPageQuery } from 'mhz-helpers';

import { API_MANAGER } from '@/manager/constants';

export function getManagers(query: Ref<IPageQuery | number>) {
  async function fn(): Promise<{ data: IManager[]; total: number }> {
    const params =
      typeof query.value === 'number'
        ? { page: query.value }
        : {
            page: query.value.page || 1,
            sort: query.value.sort.value,
            dir: query.value.sort.isAsc === false ? 'desc' : 'asc',
          };

    const { data } = await api.get(API_MANAGER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_MANAGER, query], queryFn: fn });
}

export function getManager(id?: ComputedRef<string | string[]>) {
  async function fn(): Promise<IManager | null> {
    if (!id?.value) return null;

    const { data } = await api.get(`${API_MANAGER}/${id.value}`);

    return data;
  }

  return useQuery({ queryKey: [API_MANAGER, id], queryFn: fn, refetchOnMount: true });
}

export function postManager(options: object) {
  async function fn(formData: IManager) {
    await api.post(API_MANAGER, formData);
  }

  return useMutation({ mutationKey: [API_MANAGER], mutationFn: fn, ...options });
}

export function updateManager(id: ComputedRef<string | undefined>, options: object) {
  async function fn(formData: IManager) {
    await api.patch(`${API_MANAGER}/${id.value}`, formData);
  }

  return useMutation({ mutationKey: [API_MANAGER, id], mutationFn: fn, ...options });
}

export function deleteManager(options: object) {
  async function fn(id?: string) {
    await api.delete(`${API_MANAGER}/${id}`);
  }

  return useMutation({ mutationKey: [API_MANAGER], mutationFn: fn, ...options });
}
