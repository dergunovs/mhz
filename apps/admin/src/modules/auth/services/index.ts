import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
