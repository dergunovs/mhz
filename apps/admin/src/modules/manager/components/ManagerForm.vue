<template>
  <form @submit.prevent="props.manager?._id ? update() : submit()" :class="$style.form">
    <UiField label="First name" isRequired :error="error('firstName')">
      <UiInput v-model="formData.firstName" isFocus />
    </UiField>

    <UiField label="Last name" isRequired :error="error('lastName')">
      <UiInput v-model="formData.lastName" />
    </UiField>

    <UiField label="Email" isRequired :error="error('email')">
      <UiInput v-model="formData.email" />
    </UiField>

    <UiField v-if="!props.manager?._id" label="Password" isRequired :error="error('password')">
      <UiInput v-model="formData.password" />
    </UiField>

    <FormButtons :id="props.manager?._id" :isLoading="isLoadingPost || isLoadingUpdate" @delete="handleDelete" />
  </form>
</template>

<script setup lang="ts">
import { ref, computed, onMounted } from 'vue';
import { useRouter } from 'vue-router';

import { useQueryClient } from '@tanstack/vue-query';

import { UiField, UiInput, toast } from 'mhz-ui';
import { IManager } from 'mhz-types';
import { useValidator, required } from 'mhz-validate';
import { clone } from 'mhz-helpers';

import FormButtons from '@/common/components/FormButtons.vue';

import { API_MANAGER, URL_MANAGER } from '@/manager/constants';
import { postManager, updateManager, deleteManager } from '@/manager/services';

interface IProps {
  manager?: IManager;
}

const props = defineProps<IProps>();

const queryClient = useQueryClient();

const router = useRouter();

const formData = ref<IManager>({
  firstName: '',
  lastName: '',
  email: '',
  password: '',
});

const managerId = computed(() => props.manager?._id);

const { mutate: mutatePost, isLoading: isLoadingPost } = postManager({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_MANAGER, 1], exact: true });
    toast.success('Manager added');
    router.push(URL_MANAGER);
  },
});

const { mutate: mutateUpdate, isLoading: isLoadingUpdate } = updateManager(managerId, {
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_MANAGER, props.manager?._id], exact: true });
    await queryClient.refetchQueries({ queryKey: [API_MANAGER, 1], exact: true });
    toast.success('Manager updated');
  },
});

const { mutate: mutateDelete } = deleteManager({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_MANAGER, 1], exact: true });
    toast.success('Manager deleted');
    router.push(URL_MANAGER);
  },
});

const rules = computed(() => {
  return {
    firstName: required,
    lastName: required,
    email: required,
    password: !props.manager?._id && required,
  };
});

const { error, isValid } = useValidator(formData, rules);

function submit() {
  if (isValid()) mutatePost(formData.value);
}

function update() {
  if (isValid()) mutateUpdate(formData.value);
}

function handleDelete(id: string) {
  mutateDelete(id);
}

onMounted(() => {
  if (props.manager) formData.value = clone(props.manager);
});
</script>

<style module lang="scss">
.form {
  display: flex;
  flex-direction: column;
  gap: 24px;
  align-items: flex-start;
}
</style>
