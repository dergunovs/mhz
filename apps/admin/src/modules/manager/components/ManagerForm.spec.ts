import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount, flushPromises } from '@vue/test-utils';
import { API_MANAGER, IManager, IBaseReply } from 'mhz-contracts';
import { toast } from 'mhz-ui';
import * as helpers from 'mhz-helpers';
import { dataTest } from 'mhz-helpers';

import ManagerForm from './ManagerForm.vue';

import { wrapperFactory, mockMutationReply, router } from '@/common/test';
import { MANAGERS } from '@/manager/fixtures';
import { URL_MANAGER } from '@/manager/constants';
import * as managerServices from '@/manager/services';

let onSuccessPost: () => void;
const spyMutatePost = vi.fn();

vi.spyOn(managerServices, 'postManager').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessPost = options.onSuccess;

  return mockMutationReply<IBaseReply, IManager>(spyMutatePost);
});

let onSuccessUpdate: () => void;
const spyMutateUpdate = vi.fn();

vi.spyOn(managerServices, 'updateManager').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessUpdate = options.onSuccess;

  return mockMutationReply<IBaseReply, IManager>(spyMutateUpdate);
});

let onSuccessDelete: () => void;
const spyMutateDelete = vi.fn();

vi.spyOn(managerServices, 'deleteManager').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessDelete = options.onSuccess;

  return mockMutationReply<IBaseReply, string | undefined>(spyMutateDelete);
});

const spyRefetchQueries = vi.fn();
const spyRemoveQueries = vi.fn();

vi.spyOn(helpers, 'useQueryClient').mockReturnValue({
  refetchQueries: spyRefetchQueries,
  removeQueries: spyRemoveQueries,
} as unknown as helpers.QueryClient);

const spyRouterPush = vi.spyOn(router, 'push');
const spyToastSuccess = vi.spyOn(toast, 'success');

const MANAGER = MANAGERS.data[0];
const QUERY_KEY = { queryKey: [API_MANAGER] };

const managerForm = dataTest('manager-form');
const managerFormFirstName = dataTest('manager-form-first-name');
const managerFormLastName = dataTest('manager-form-last-name');
const managerFormEmail = dataTest('manager-form-email');
const managerFormPassword = dataTest('manager-form-password');
const managerFormButtons = dataTest('manager-form-buttons');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(ManagerForm, {
    props: {
      manager: MANAGER,
    },
  });
});

enableAutoUnmount(afterEach);

describe('ManagerForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ManagerForm)).toBeTruthy();
  });

  it('shows manager content in fields if props is passed', async () => {
    expect(wrapper.findComponent(managerFormFirstName).attributes('modelvalue')).toBe(MANAGER.firstName);
    expect(wrapper.findComponent(managerFormLastName).attributes('modelvalue')).toBe(MANAGER.lastName);
    expect(wrapper.findComponent(managerFormEmail).attributes('modelvalue')).toBe(MANAGER.email);
  });

  it('updates manager', async () => {
    expect(spyMutateUpdate).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    const NEW_MANAGER_EMAIL = 'new@email.com';

    await wrapper.findComponent(managerFormEmail).setValue(NEW_MANAGER_EMAIL);

    await wrapper.find(managerForm).trigger('submit');

    expect(spyMutateUpdate).toBeCalledTimes(1);
    expect(spyMutateUpdate).toBeCalledWith({ ...MANAGER, email: NEW_MANAGER_EMAIL });

    onSuccessUpdate();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('deletes manager', async () => {
    expect(spyMutateDelete).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    wrapper.findComponent<DefineComponent>(managerFormButtons).vm.$emit('delete', MANAGER._id);

    expect(spyMutateDelete).toBeCalledTimes(1);
    expect(spyMutateDelete).toBeCalledWith(MANAGER._id);

    onSuccessDelete();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith(QUERY_KEY);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_MANAGER);
  });

  it('creates manager', async () => {
    wrapper.unmount();

    expect(spyMutatePost).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    const wrapperWithoutManager = wrapperFactory(ManagerForm, {});

    const NEW_MANAGER = {
      firstName: 'Petr',
      lastName: 'Alekseev',
      email: 'new@mail.ru',
      password: 'qwerty',
    };

    await wrapperWithoutManager.findComponent(managerFormFirstName).setValue(NEW_MANAGER.firstName);
    await wrapperWithoutManager.findComponent(managerFormLastName).setValue(NEW_MANAGER.lastName);
    await wrapperWithoutManager.findComponent(managerFormEmail).setValue(NEW_MANAGER.email);
    await wrapperWithoutManager.findComponent(managerFormPassword).setValue(NEW_MANAGER.password);

    await wrapperWithoutManager.find(managerForm).trigger('submit');

    expect(spyMutatePost).toBeCalledTimes(1);
    expect(spyMutatePost).toBeCalledWith(NEW_MANAGER);

    onSuccessPost();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_MANAGER);
  });
});
