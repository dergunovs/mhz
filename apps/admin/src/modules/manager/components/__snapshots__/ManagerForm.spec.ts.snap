// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManagerForm > matches snapshot 1`] = `
"<form class="_form_72252e" data-test="manager-form">
  <fieldset label="First name" isrequired=""><input type="text" modelvalue="Alexandr" isfocus="" data-test="manager-form-first-name"></fieldset>
  <fieldset label="Last name" isrequired=""><input type="text" modelvalue="Petrov" data-test="manager-form-last-name"></fieldset>
  <fieldset label="Email" isrequired=""><input type="text" modelvalue="a@mail.ru" data-test="manager-form-email"></fieldset>
  <!--v-if-->
  <form-buttons-stub id="64f19fc5677cb758200ab511" data-test="manager-form-buttons"></form-buttons-stub>
</form>"
`;
